shimmingtoolbox/shim/realtime_zshim.py

```diff
@@ -113,21 +122,88 @@ def realtime_zshim(nii_fieldmap, nii_anat, pmu, json_fmap, nii_mask_anat=None, p
     static = np.array([np.zeros_like(fieldmap[:, :, :, 0]),
                        np.zeros_like(fieldmap[:, :, :, 0]),
                        np.zeros_like(fieldmap[:, :, :, 0])])
-    # TODO fix progress bar not showing up
-    progress_bar = st_progress_bar(fieldmap[..., 0].size * 3, desc="Fitting", ascii=False)
+
+
+    p = acq_pressures - mean_p
+    nVoxels = nx * ny * nz
+    nVoxPerSlice = nx * ny
+
+    I = sparse.identity(nVoxels,format='csc')
+
+    # lamda[0]: regularization parameter for the static field component
+    # lamda[1]: regularization parameter for the RIRO component
+    lamda = np.array([20, 30000]) # regularization parameters selected by trial and error
+    #lamda = np.array([0, 0]) # regularization parameters selected by trial and error
+
+    # first order finite difference operator: D
+    # 1st dimension
+    e = np.ones(nx)
+    diags = np.array([-1, 1])
+    Dx = spdiags([-e, e], diags, nx, nx).toarray()
+
+    while Dx.shape[1] < nVoxels :
+        Dx = block_diag(Dx, Dx)
+    Dx = Dx[0:nVoxels,0:nVoxels]     
+
+    # 2nd dimension
+    e = np.ones(nVoxPerSlice)
+    diags = np.array([-nx, nx])
+    Dy = spdiags([-e, e], diags, nVoxPerSlice, nVoxPerSlice).toarray() 
+
+    while Dy.shape[1] < nVoxels :
+        Dy = block_diag(Dy, Dy)  
+    Dy = Dy[0:nVoxels,0:nVoxels]
+
+    # 3rd dimension
+    e = np.ones(nVoxels)
+    diags = np.array([-nVoxPerSlice, nVoxPerSlice])
+    Dz = spdiags([-e, e], diags, nVoxels, nVoxels).toarray() 
+
+    Dx = Dx/(2*delx) 
+    Dy = Dy/(2*dely) 
+    Dz = Dz/(2*delz) 
+    L = coo_matrix(Dx + Dy + Dz)
```

Comment:
I need to spend more time on this (will do monday) but I think the error is introduced around this line, more specifically `Dy` seems to create the pattern in the image. (I tried only running with `Dy` instead of `Dx+Dy+Dz` and the error is similar). I would revisit the calculation of those terms (or how they are used). I don't see any obvious mistake though.

Reply:
oooh, good find! The error is gone without Dy. I'll narrow down on Dy then ...

Reply:
Ok, I've figured out what's going on here. 

Previously I defined the following:
`nx, ny, nz, nt = nii_fieldmap.shape`, which returns **64 x 96 x 1 x 60**
numpy.shape presumably gives dimensions in the following order: **(metaframes, frames, rows, columns)**

In the case of `nii_fieldmap.shape`, that is not the order (I do not know why, and I haven't found any documentation giving information on this). I had _assumed_ the order was: **(rows, columns, frames, metaframes)**.

In the matlab code [(createdifferenceoperators.m)](https://github.com/shimming-toolbox/shimming-toolbox-matlab/blob/eao/SpatialRegularization/misc/createdifferenceoperators.m), it is stated that:
`Dx, Dy, & Dz that act as partial differential operators in row, column, and slice directions respectively.`

In matlab, the image matrix dimensions are: **96 x 64 x 1 x 1 x 60**, which corresponds to **rows x columns x slices x echoes x frames**. 

So, it looks like the matrix dimensions are flipped for python vs. matlab (?) 

What I've now done is changed the following:
`nCol, nRow, nSlice, nt = nii_fieldmap.shape`. This goes against the assumption I made above (that  `nii_fieldmap.shape` returns  **(rows, columns, frames, metaframes)**. 

The code now outputs static and riro correction data that make more sense (no more artifact), but don't exactly match the matlab-code output. However, the least-squares algorithm is not the same, so we can't expect the results to be (making it hard to validate whether or not the indexing is correct now).

Reply:
> So, it looks like the matrix dimensions are flipped for python vs. matlab (?)

Either Numpy or Nibabel does make the orientation different from Matlab as you have seen:
- Matlab: 96 x 64 x 1 x 1 x 60
- Python: 64 x 96 x 1 x 60

I used the default for nibabel and numpy and they return nCol, nRow, nSlice, nt and I use different variables for the different echoes. The indexes for either the column or row are also indexed the other way (i.e. One of these is true:  Row[0] in Python == Row[96] in Matlab OR Col[0] in Python == Col[64] in Matlab but I never remember which one is true). I like to think of it as if the first 2 dimensions got rotated. (This is why all plots in python are rotated from the plots in matlab.